tools/Voron.Recovery/RecoveredDatabaseCreator.cs

```diff
@@ -0,0 +1,433 @@
+{
+    public class RecoveredDatabaseCreator : IDisposable
+    {
+        public static RecoveredDatabaseCreator RecoveredDbTools(DocumentDatabase documentDatabase, string recoverySession, Logger logger) => new RecoveredDatabaseCreator(documentDatabase, recoverySession, logger);
+
+        private readonly DatabaseDestination _databaseDestination;
+        private readonly SmugglerResult _results;
+        private readonly DocumentsOperationContext _context;
+        private readonly DocumentDatabase _database;
+        private readonly IDisposable _contextDisposal;
+        private readonly Logger _logger;
+        private readonly string _logDocId;
+        private readonly string _orphanAttachmentsDocIdPrefix;
+        private readonly string _orphanCountersDocIdPrefix;
+        private readonly Slice _attachmentsSlice;
+        private readonly ByteStringContext _byteStringContext;
+        private readonly string _recoveryLogCollection;
+        private string GetOrphanAttachmentDocId(string hash) => $"{_orphanAttachmentsDocIdPrefix}/{hash}";
+
+        private RecoveredDatabaseCreator(DocumentDatabase documentDatabase, string recoverySession, Logger logger)
+        {
+            _logger = logger;
+            _recoveryLogCollection = $"RecoveryLog-{recoverySession}";
+            _orphanAttachmentsDocIdPrefix = $"OrphanAttachments/{recoverySession}";
+            _orphanCountersDocIdPrefix = $"OrphanCounters/{recoverySession}";
+            _logDocId = $"Log/{recoverySession}";
+            _database = documentDatabase;
+            _databaseDestination = new DatabaseDestination(documentDatabase);
+            _results = new SmugglerResult();
+            _databaseDestination.Initialize(new DatabaseSmugglerOptionsServerSide(), _results, ServerVersion.Build);
+            _contextDisposal = _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+            _byteStringContext = new ByteStringContext(SharedMultipleUseFlag.None);
+            Slice.From(_byteStringContext, "Attachments", ByteStringType.Immutable, out _attachmentsSlice);
+
+            using (var tx = _context.OpenWriteTransaction())
+            {
+                using (var doc = _context.ReadObject(
+                    new DynamicJsonValue
+                    {
+                        ["RecoverySession"] = recoverySession,
+                        ["RecoveryStarted"] = DateTime.Now,
+                        [Constants.Documents.Metadata.Key] = new DynamicJsonValue {[Constants.Documents.Metadata.Collection] = _recoveryLogCollection}
+                    }, _logDocId, BlittableJsonDocumentBuilder.UsageMode.ToDisk)
+                )
+                {
+                    _database.DocumentsStorage.Put(_context, _logDocId, null, doc);
+                }
+                tx.Commit();
+            }
+        }
+
+        public void WriteDocument(Document document)
+        {
+            var actions = _databaseDestination.Documents();
```

Comment:
using

Reply:
no using here because sometimes I do want manually dispose it (and then enforce commit)

Reply:
But you aren't doing any disposal internally at all, no?

Reply:
I do: https://github.com/ravendb/ravendb/pull/10266/files#diff-54908524c9c84febb0b2eaa52e9d26c1R277

Reply:
Why are you enforcing commit? What is the idea behind it?

Reply:
To use different transactions.

I am using DatabaseDestination (and not direct PUT command) in order to be compatible with what import does for instance. And then I want to commit (`actions.Dispose()`) and add another transaction _only if needed_. 

I cannot do everything in a single transaction AND use DatabaseDestination/TxMerger.